nixos/modules/services/networking/soju.nix

```diff
@@ -0,0 +1,126 @@
+  listenCfg = concatMapStringsSep "\n" (l: "listen ${l}") cfg.listen;
+  tlsCfg = optionalString (cfg.tlsCertificate != null)
+    "tls ${cfg.tlsCertificate} ${cfg.tlsCertificateKey}";
+  logCfg = optionalString (cfg.logDirectory != null)
+    "log fs ${cfg.logDirectory}";
+
+  configFile = pkgs.writeText "soju.conf" ''
+    ${listenCfg}
+    hostname ${cfg.hostName}
+    ${tlsCfg}
+    db sqlite3 ${cfg.databasePath}
+    ${logCfg}
+    http-origin ${concatStringsSep " " cfg.httpOrigins}
+    accept-proxy-ip ${concatStringsSep " " cfg.acceptProxyIP}
+
+    ${cfg.extraConfig}
+  '';
+in
+{
+  ###### interface
+
+  options.services.soju = {
+    enable = mkEnableOption "soju";
+
+    listen = mkOption {
+      type = types.listOf types.str;
+      default = [ ":6697" ];
+      description = ''
+        Where soju should listen for incoming connections. See the
+        <literal>listen</literal> directive in
+        <citerefentry><refentrytitle>soju</refentrytitle>
+        <manvolnum>1</manvolnum></citerefentry>.
+      '';
+    };
+
+    hostName = mkOption {
+      type = types.str;
+      default = config.networking.hostName;
+      description = "Server hostname.";
+    };
+
+    tlsCertificate = mkOption {
+      type = types.nullOr types.path;
+      example = "/var/host.cert";
+      description = "Path to server TLS certificate.";
+    };
+
+    tlsCertificateKey = mkOption {
+      type = types.nullOr types.path;
+      example = "/var/host.key";
+      description = "Path to server TLS certificate key.";
+    };
+
+    databasePath = mkOption {
+      type = types.path;
+      default = "/var/lib/soju/soju.db";
+      description = "Path to the SQLite database.";
+    };
+
+    logDirectory = mkOption {
+      type = types.nullOr types.path;
+      example = "/var/lib/soju/logs";
+      description = "Path to the bouncer logs root directory.";
+    };
+
+    httpOrigins = mkOption {
+      type = types.listOf types.str;
+      default = [];
+      description = ''
+        List of allowed HTTP origins for WebSocket listeners. The parameters are
+        interpreted as shell patterns, see
+        <citerefentry><refentrytitle>glob</refentrytitle>
+        <manvolnum>7</manvolnum></citerefentry>.
+      '';
+    };
+
+    acceptProxyIP = mkOption {
+      type = types.listOf types.str;
+      default = [];
+      description = ''
+        Allow the specified IPs to act as a proxy. Proxys have the ability to
+        overwrite the remote and local connection addresses (via the X-Forwarded-\*
+        HTTP header fields). The special name "localhost" accepts the loopback
+        addresses 127.0.0.0/8 and ::1/128. By default, all IPs are rejected.
+      '';
+    };
+
+    extraConfig = mkOption {
+      type = types.lines;
+      default = "";
+      description = "Lines added verbatim to the configuration file.";
+    };
+  };
+
+  ###### implementation
+
+  config = mkIf cfg.enable {
+    users.users.soju = {
```

Comment:
Did you consider `DynamicUser` instead?

Reply:
How would I add such a user to the `acme` group?

Reply:
With the [SupplementaryGroups](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#SupplementaryGroups=) directive.

Reply:
Ok, maybe I am a bit biased here because I'm not a big fan of systemd-ifying everything, but isn't this a rather non-standard approach? Adding a user to the `acme` group via `extraGroups` is simple, whereas I wouldn't expect the average NixOS user to know about SupplementaryGroups.

Reply:
I was just trying to offer some helpful suggestions based on https://github.com/NixOS/rfcs/blob/master/rfcs/0052-dynamic-ids.md.

I think having the user set `extraGroups` sounds time consuming as well... I would suggest using the `useACMEHost` pattern you will find in NixOS to take care of it for the sysadmin! Let me know if you need any advice or help on that.

Reply:
Thanks for pointing me to the RFC! Having systemd take care of all the permissions without assigning static uids does sound appealing. I don't know what `useACMEHost` should do though.

Reply:
`useACMEHost` should work together with our `security.acme` module to provide ssl certificates for your module.

Take a look at [this](https://github.com/NixOS/nixpkgs/blob/nixos-21.05/nixos/modules/services/web-servers/pomerium.nix) as a rough guide.

If you want to implement this and need help just ping me.

Reply:
Looking at the pomerium implementation, it seems like a lot of work to accomplish such a simple thing. I would actually prefer `systemd.services.soju.serviceConfig.SupplementaryGroups` to that.

Reply:
Switched to `DynamicUser`.